### PR TITLE
Gender refactor

### PR DESCRIPTION
**Pull request by AI0867**
unit_gender is now no longer a member of unit_race.

I'm not entirely sure the full set of commits are the right direction, but part of the branch can also be used. `MAKE_ENUM` is another option.

### CODE REVIEW

**Comment by Vultraz** [comment]
Looks alright, but please format the new unit_gender.cpp file correctly. Ie, move the leading function brace to its own line and add a newline between functions. Also drop the commented out `static` qualifiers.

**Comment by AI0867** [comment]
> Looks alright, but please format the new unit_gender.cpp file correctly. Ie, move the leading function brace to its own line[...]

Is that the official coding style now?
> Also drop the commented out static qualifiers.

They are there as an indication that these are class methods rather than instance methods.

**Comment by Vultraz** [comment]
> Is that the official coding style now?
Yes

> They are there as an indication that these are class methods rather than instance methods.
That's already evident enough.

**Comment by GregoryLundberg** [comment]
Appveyor failure due to out-of-date project files.

File: src/units/gender.hpp [hunk 1]
```diff
@@ -16,11 +16,42 @@
...
+public:
+	static const unit_gender& male();
+	static const unit_gender& female();
+	static constexpr int num_genders(){
+		return 2;
+	}
+
+	static const unit_gender* from_string(const std::string&);
+	static const unit_gender& from_string(const std::string&, const unit_gender& fallback);
+	static const unit_gender* from_int(int);
```
Comment (CelticMinstrel) [inline src/units/gender.hpp]: I don't think I really like the idea of passing around pointers to genders; I'd rather it be a lightweight type that can be cheaply copied.

File: src/units/gender.cpp [hunk 2]
```diff
@@ -17,26 +17,79 @@
 
 #include <ostream>
 
+namespace {
+	/// Standard string id (not translatable) for FEMALE
+	const std::string s_female("female");
+	/// Standard string id (not translatable) for MALE
+	const std::string s_male("male");
```
Comment (CelticMinstrel) [inline src/units/gender.cpp]: I'm pretty sure these need to be marked translatable. Obviously you can't use the `_()` syntax, but there's an alternate syntax that marks it as translatable for gettext without implying an actual lookup (you'd then look it up later, probably via the `t_string` constructor).

File: src/units/gender.cpp [hunk 3]
```diff
@@ -17,26 +17,79 @@
...
-		return unit_gender::MALE;
-	} else if ( str == unit_race::s_female ) {
-		return unit_gender::FEMALE;
+const char* unit_gender::gender_string(const char* male_string, const char* female_string) const {
+	return *this == female() ? female_string : male_string;
+}
+const std::string& unit_gender::gender_string(const std::string& male_string, const std::string& female_string) const {
+	return *this == female() ? female_string : male_string;
+}
+const t_string& unit_gender::gender_string(const t_string& male_string, const t_string& female_string) const {
```
Comment (CelticMinstrel) [inline src/units/gender.cpp]: I wonder what the need is for three functions here, given that `t_string` can be constructed from both of the others. Plus, these would normally be used on translatable strings anyway. I don't see any reason to use them on non-translatable strings?

File: src/units/gender.cpp [hunk 4]
```diff
@@ -63,11 +63,8 @@ unit_gender::unit_gender(int index, const std::string& name)
 	const unit_gender* gender = from_string(str_gender);
 	return gender ? *gender : fallback;
 }
-/*static*/ const unit_gender* unit_gender::from_int(int index){
-	assert(index < unit_gender::num_genders());
-	auto gender = gender_map[index];
-	assert(gender);
-	return gender;
+/*static*/ gender_list unit_gender::genders(){
+	return {};
```
Comment (CelticMinstrel) [inline src/units/gender.cpp]: ...what? This is just weird, why are there no genders...

File: src/units/gender.hpp [hunk 5]
```diff
@@ -19,6 +19,23 @@
 class t_string;
 class unit_gender;
 
+class gender_list {
```
Comment (CelticMinstrel) [inline src/units/gender.hpp]: This seems really overcomplicated for no real reason...

File: src/units/gender.cpp (hunk 2 above)
Comment (AI0867) [reply src/units/gender.cpp]: These were always untranslatable, see current `src/unit/race.cpp`. They are simply identifiers, used for WML keys and such. The actual translatable versions are in different locations. Refactoring that might be a good idea, but is out of scope of this PR.

File: src/units/gender.cpp (hunk 4 above)
Comment (AI0867) [reply src/units/gender.cpp]: The `gender_list` class has no members, and only a default constructor. Its function is to provide `begin()` and `end()` methods for iteration. This method simply constructs a `gender_list`.

File: src/units/gender.cpp (hunk 3 above)
Comment (AI0867) [reply src/units/gender.cpp]: The caller may want a `const char*` type string, and while you could extract that from a `t_string`, the freshly constructed `t_string` would be a temporary, and you'd have a pointer to a member of a member of a temporary, which is is just begging for undefined behavior. We once had similar problems with configs that returned `t_string`. (before `config::attribute_value`)

File: src/units/gender.hpp (hunk 1 above)
Comment (AI0867) [reply src/units/gender.hpp]: With pointers, it's just as cheap as the original version, and lacks the special `NUM_GENDERS` value.

Other options would be `MAKE_ENUM` or just the first commit of this branch.

File: src/units/gender.hpp (hunk 5 above)
Comment (AI0867) [reply src/units/gender.hpp]: It's about as simple as you can make an iterator system, it seems.
I added it to get rid of the `from_int` lookup, and convert integer iteration to range-for.

File: src/units/gender.hpp [hunk 6]
```diff
@@ -16,11 +16,42 @@
 
 #include <iosfwd>
 
-enum class unit_gender {
-	MALE, FEMALE, NUM_GENDERS
+class t_string;
+class unit_gender;
```
[the remark this answers is not in the document]
Comment (AI0867) [reply src/units/gender.hpp]: It's needed so the iterator can refer to unit_gender.
Swapping the order of the classes would simply move the problem around.

**Comment by AI0867** [comment]
Well, the `nullptr`s represent what was previously represented by a value of `NUM_GENDERS`. Every function that accepts or returns a pointer does so because not passing a gender is permitted. Every function that guarantees or requires a valid gender uses references instead.

File: src/units/gender.cpp (hunk 2 above)
Comment (CelticMinstrel) [reply src/units/gender.cpp]: Ah, okay.

File: src/units/gender.cpp (hunk 4 above)
Comment (CelticMinstrel) [reply src/units/gender.cpp]: I find this syntax confusing. I'd rather see a constructor call here.

File: src/units/gender.hpp (hunk 5 above)
Comment (CelticMinstrel) [reply src/units/gender.hpp]: Yeah, uhh... it seems overcomplicated _because_ it's an iterator system for a set of _two fixed values_.

File: src/units/gender.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (CelticMinstrel) [reply src/units/gender.hpp]: Hmm, maybe that was added in a commit after where I saw this line, then. If I recall correctly, what I saw that prompted this comment was literally:

```
class unit_gender;

class unit_gender
{
```

File: src/menu_events.cpp [hunk 7]
```diff
@@ -701,14 +701,14 @@ void create_and_place(game_display&,
...
-		unit_race::GENDER gender = unit_race::NUM_GENDERS)
+		const unit_gender* gender = nullptr)
 {
 	synced_context::run_and_throw("debug_create_unit",
 		config {
 			"x", loc.wml_x(),
 			"y", loc.wml_y(),
 			"type", u_type.id(),
-			"gender", gender_string(gender),
+			"gender", gender ? gender->str() : unit_gender::male().str(),
```
Comment (CelticMinstrel) [inline src/menu_events.cpp]: This sort of thing also bugs me a little. Maybe it's not exactly best practice, but couldn't we do something like this?

```

std::string unit_gender::str() {
    if(this == nullptr) return unit_gender::male().str();
    // normal implementation of str() here
}
```

File: src/units/gender.cpp [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	const std::string s_male("male");
+	/// Storage for gender map
+	std::vector<const unit_gender*> gender_map = std::vector<const unit_gender*>(unit_gender::num_genders(), nullptr);
+	// Populate the map
+	auto & m_ = unit_gender::male();
+	auto & f_ = unit_gender::female();
+
+	// C++17 as_const
+	template<class T>
+	T const& constant(T& v){ return v; }
```
Comment (CelticMinstrel) [inline src/units/gender.cpp]: I don't see this being used anywhere?

**Comment by CelticMinstrel** [comment]
If `nullptr` is technically a valid gender value in some places, I wonder if we should make the class null-safe (ie, so that it doesn't crash if you call member functions on a null pointer).

**Comment by AI0867** [comment]
> If `nullptr` is technically a valid gender value in some places, I wonder if we should make the class null-safe (ie, so that it doesn't crash if you call member functions on a null pointer).

While this sounds like a decent idea, and we've had code in wesnoth that behaved like this, calling a member of a `nullptr`, even a non-virtual one, is undefined behavior.

File: src/menu_events.cpp (hunk 7 above)
Comment (gfgtdf) [reply src/menu_events.cpp]: we can't. Calling a memeber function on a nullptr is undefioned behviour, in particular the compiler might simply assume 'this != nullptr' and optimize that check out.

**Comment by spixi** [comment]
So, what about non-binary genders? Who says, that a race can only have two genders? I feel, it would be better to push the gender definition to WML/Lua handle them generically in C++.

**Comment by AI0867** [comment]
Binary genders are hardcoded into a lot of places (just see all the ternary operators in the diff), and changing that is beyond the scope of this PR.

Regarding the PR itself, it can be split into three parts:

1. Moving gender out of unit_race (0566c3c0cf8d53dc6fba410c9947e994f96ff34b). This means the unit_race class no longer needs to be included just to manipulate genders.
2. Turning unit_gender into a class, so string conversions and the like can be centralized. (275edd02737584996c24bdc470752233d6cae8ae, dca2bef707432bfb6d18e82df25d79ac5cc4ce11, 6f6808280fd6413f522d77f567d5d4d7a73f8ede, b33c10d70bd28014787d45a3fffeadfc6d7b1cd1)
3. Changing the integer-conversion into an iterator. This is probably the most controversial. (c986fd102e31f53682b126bb2edc9a614082d230)

And then there are style commits: e92d9ca6d9cb784e8392b98e01162cb782806095, f830833a7a5d68ed2b45e3fdb6bd3143b1f6ca21

Part of it can probably be cherry-picked.

**Comment by Pentarctagon** [comment]
@AI0867 are you going to continue working on this PR?

**Comment by AI0867** [comment]
There isn't really any more work to be done, just deciding which parts of it we actually want to merge.

If we want to support more than 2 genders, additional work would need to be done, but that can be built on top of this.

If the PR needs to be updated to fix merge conflicts, I update the parts that we actually want.

File: src/units/race.hpp [hunk 9]
```diff
@@ -31,12 +31,12 @@ class unit_race
 	const config& get_cfg() const { return cfg_; }
 	const std::string& id() const { return id_; }
 	const std::string& editor_icon() const { return icon_; }
-	const t_string& name(GENDER gender=MALE) const { return name_[gender]; }
+	const t_string& name(const unit_gender* gender = nullptr) const;
```
Comment (stevecotton) [inline src/units/race.hpp:34]: I think many functions should stop providing default values for their arguments. For example, by removing the default here we find that the single function which depends on it is the implementation of the Lua API; thus there's a Lua API which is assuming the gender.

**Comment by stevecotton** [comment]
Closing this, because I think it should now be implemented using #6470 's `enum_base` class instead. This PR will be a useful reference to use as a checklist of things that might be missed when someone does an `enum_base` version, but I'd expect any new implementation to start from the current source instead of resolving merge conflicts from when this forked.

I'm not volunteering to do the task, just considering how it would be done.

**Pull request closed without merging**